### PR TITLE
Dispatch event before Report's query execution for plugins to alter query

### PR DESCRIPTION
**Pull request by scottshipman**
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | NO
| New feature? | YES
| Automated tests included? | NO
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Creates a ReportQueryEvent class and registers a REPORT_QUERY_PRE_EXECUTE event that is dispatched prior to getReportData's query execution so that a plugin can alter the query.  

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. 
2. 

#### Steps to test this PR:
1. render a report view. 
2. 

#### List deprecations along with the new alternative:
1. 
2. 

#### List backwards compatibility breaks:
1. 
2.

### CODE REVIEW

File: app/bundles/ReportBundle/Event/ReportQueryEvent.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
+<?php
+
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Comment (Maxell92) [inline app/bundles/ReportBundle/Event/ReportQueryEvent.php]: Could you change the year to 2018, please?

File: app/bundles/ReportBundle/Model/ReportModel.php [hunk 2]
```diff
@@ -239,15 +240,15 @@ public function buildAvailableReports($context)
                 $data[$context]['graphs'] = &$data['all']['graphs'][$context];
             } else {
                 //build them
-                $eventContext = ($context == 'all') ? '' : $context;
+                $eventContext = ('all' == $context) ? '' : $context;
```
Comment (Maxell92) [inline app/bundles/ReportBundle/Model/ReportModel.php]: This seems to me that you have some custom CS Fixer or any other tool set?
Could you provide this to the community if so?

Comment (scottshipman) [reply app/bundles/ReportBundle/Model/ReportModel.php]: I was using the mautic provided config file, contents pasted below. I noticed it tends to change things that mautic core has not been validated against yet. Is there a Mautic policy for php-cs-fixer, like when to or not to use it?

```
<?php

$finder = PhpCsFixer\Finder::create()
    ->in(__DIR__.'/app/bundles')
    ->in(__DIR__.'/app/config')
    ->in(__DIR__.'/app/middlewares')
    ->in(__DIR__.'/app/migrations')
    ->in(__DIR__.'/plugins');

return PhpCsFixer\Config::create()
    ->setRules([
        '@Symfony' => true,
        'binary_operator_spaces' => [
            'align_double_arrow' => true,
            'align_equals' => true
        ],
        'ordered_imports' => true,
        'array_syntax' => [
            'syntax' => 'short'
        ],
    ])
    ->setFinder($finder);
```

Comment (Maxell92) [reply app/bundles/ReportBundle/Model/ReportModel.php]: This came from Symfony Coding standards, so Mautic should use it. However CS Fixer does not touch it for many devs. It is not checked on Travis too.

**Comment by escopecz** [comment]
@scottshipman it's failing tests for PHP 5.6. Please take a look. 

https://travis-ci.org/mautic/mautic/jobs/375996613#L671

**Comment by scottshipman** [comment]
All tests passed.

File: app/bundles/ReportBundle/Event/ReportQueryEvent.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+     * @var int
+     */
+    private $totalResults = 0;
+
+    /**
+     * ReportDataEvent constructor.
+     *
+     * @param Report       $report
+     * @param QueryBuilder $query
+     * @param              $totalResults
```
Comment (escopecz) [inline app/bundles/ReportBundle/Event/ReportQueryEvent.php]: Can you add `int` as a type hint here please?

**Pull request merged**